File: frontend/src/api/crypto.js
const BASE_URL = "https://crypto-wave-8j9w.onrender.com/api/crypto";
console.log("API URL:", import.meta.env.VITE_API_URL);
// Umumiy so'rov yuboruvchi funksiya
async function request(endpoint, body) {
  const response = await fetch(`${BASE_URL}/${endpoint}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });

  const data = await response.json();

  // Backend xato qaytarsa
  if (!response.ok) {
    throw new Error(data.detail || "Xatolik yuz berdi");
  }

  return data;
}

// --- Caesar ---
export const caesarEncode = (text, key) =>
  request("caesar/encode", { text, key });

export const caesarDecode = (text, key) =>
  request("caesar/decode", { text, key });

// --- Base64 ---
export const base64Encode = (text) =>
  request("base64/encode", { text });

export const base64Decode = (text) =>
  request("base64/decode", { text });

// --- AES ---
export const aesEncode = (text, key) =>
  request("aes/encode", { text, key });

export const aesDecode = (text, key) =>
  request("aes/decode", { text, key });

// RSA
export const rsaGenerate = () =>
  fetch("/api/crypto/rsa/generate").then(r => r.json());

export const rsaEncode = (text, key) =>
  request("rsa/encode", { text, key });

export const rsaDecode = (text, key) =>
  request("rsa/decode", { text, key });

File: backend/app/routers/crypto.py
from fastapi import APIRouter, HTTPException
from app.schemas.crypto import EncryptRequest, DecryptRequest, CryptoResponse, RSAKeysResponse
from app.services.crypto_service import CryptoService

router = APIRouter(prefix="/api/crypto", tags=["crypto"])

# Servisni bir marta yaratamiz
service = CryptoService()


# --- BASE64 ---
@router.post("/base64/encode", response_model=CryptoResponse)
def base64_encode(request: EncryptRequest):
    try:
        result = service.base64_encode(request.text)
        return CryptoResponse(result=result, algorithm="base64", success=True)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/base64/decode", response_model=CryptoResponse)
def base64_decode(request: DecryptRequest):
    try:
        result = service.base64_decode(request.text)
        return CryptoResponse(result=result, algorithm="base64", success=True)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


# --- CAESAR ---
@router.post("/caesar/encode", response_model=CryptoResponse)
def caesar_encode(request: EncryptRequest):
    try:
        result = service.caesar_encode(request.text, request.key)
        return CryptoResponse(result=result, algorithm="caesar", success=True)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/caesar/decode", response_model=CryptoResponse)
def caesar_decode(request: DecryptRequest):
    try:
        result = service.caesar_decode(request.text, request.key)
        return CryptoResponse(result=result, algorithm="caesar", success=True)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


# --- AES ---
@router.post("/aes/encode", response_model=CryptoResponse)
def aes_encode(request: EncryptRequest):
    try:
        result = service.aes_encode(request.text, request.key)
        return CryptoResponse(result=result, algorithm="aes", success=True)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/aes/decode", response_model=CryptoResponse)
def aes_decode(request: DecryptRequest):
    try:
        result = service.aes_decode(request.text, request.key)
        return CryptoResponse(result=result, algorithm="aes", success=True)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))
    
@router.get("/rsa/generate", response_model=RSAKeysResponse)
def rsa_generate():
    try:
        keys = service.rsa_generate_keys()
        return RSAKeysResponse(
            public_key=keys["public_key"],
            private_key=keys["private_key"],
            n_bits=keys["n_bits"],
            success=True
        )
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/rsa/encode", response_model=CryptoResponse)
def rsa_encode(request: EncryptRequest):
    try:
        result = service.rsa_encode(request.text, request.key)
        return CryptoResponse(result=result, algorithm="RSA-512", success=True)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/rsa/decode", response_model=CryptoResponse)
def rsa_decode(request: DecryptRequest):
    try:
        result = service.rsa_decode(request.text, request.key)
        return CryptoResponse(result=result, algorithm="RSA-512", success=True)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))